### PR TITLE
Upgrade testing system

### PR DESCRIPTION
**Pull request by mjguynn**
This moves the Turnt-based pipeline tests to a new Python-based testing system. The explication-only tests (currently, the `scheduling` folder) are unaffected.

I started working on this for a pretty dumb reason: the Turnt scripts (for pipeline tests, not explication) use Posix utilities like `cat` which prevents them from running in the normal Windows shell. The new version indeed works in the Windows shell but it has a bunch of other benefits too.

The testing script has three modes: 
- `clean`, which is what it sounds like
- `run`, which is `build` + `cargo test`
- `build`

The `build` mode recursively searches for all `.cair` and `.ron` files in the test directory and runs caimanc on them. If they successfully compile, the generated code is emitted to `src` as a Rust module. If a Rust file exists in the same directory as the source file with the same filename, the generated code `include!(...)`s that Rust file. For example, `caiman-test/basics/external_test.cair` gets compiled to `src/external_test.rs` which contains `include!("../basics/external_test.rs")`. A major advantage this is that if the problem lies in Rust code, you can just edit the original Rust code and run `cargo test` *without* having to rerun the test script. 

All of the generated files are then "linked" into one Rust library, which you can then build and test with `cargo test`. This takes advantage of Cargo's nice testing infrastructure and output. Keeping everything in one library also makes the tests run faster since we can reuse the same WGPU instance. Also, rust-analyzer is able to understand the crate structure, which helps diagnose error/warnings/suggestions.

Instead of storing output in a separate file, you just directly check it in the testing code. Most tests use `crate::expect_returned!` to do this, which is conceptually the same as `assert_eq!` except it returns a Result instead of panicking (a panic would poison the WGPU mutex).

The tests themselves are simplified because a bunch of common boilerplate (i.e. setting up a WGPU instance) was factored out into a shared module.

### CODE REVIEW

**Comment by Checkmate50** [comment]
Hmm, I appreciate the sentiment (I am on Windows as well, and had to do a lot of fiddling to make it work on my machine without bonking Linux; I relied on cygwin though).  The main problems I have with switching, however, are twofold:
1) We end up recreating a lot of the benefits (and problems with) turnt.  This means that we end up having to add whatever is needed that Turnt already has (the most obvious being parallel testing), and end up with kind of a light-weight test suite that awkwardly in-between super-minimal tests and building a proper unit test library suite
2) We also lose any benefits of being in any ecosystem

The upsides, to be fair, seem to me to be that
1) We get control over caiman testing directly, and don't have to worry about Turnt breaking independently for some reason
2) The tests can be more customized for the caiman environment, which has pros and cons I suppose in terms of design

In short, I'm fine with it, but I'm not fully sold that we want to have a completely custom, but still light-weight test set.

**Comment by mjguynn** [comment]
To be clear the actual pipeline tests run via Cargo, so I wouldn't really say it's a custom testing system. The Python script is just an adapter that compiles all the Caiman inputs first and then sticks them in the proper folder, which is what the shell scripts inside the Turnt configs were doing anyway. As far as ecosystem goes... we're ultimately testing (generated) Rust code, and Cargo unit tests are the most popular way to do that.

I don't think we need Turnt's power for the pipeline tests. IMO the main value of the pipeline tests is making sure that we didn't break Caiman's functionality, and Cargo tests are more ergonomic for that since they're deeply integrated into Rust. (Cargo actually supports parallel testing but we don't take advantage of that right now.)

**Comment by Checkmate50** [comment]
I buy it, seems good to me then.  I admit, this is what I was tempted to do initially, and I do like having more control of course.  I'll download and fiddle with the branch when I get back from ASPLOS myself to make sure it works and will integrate well, but conceptually gets a thumbs up from me!
